controllers/snapshot/snapshot_controller.go

```diff
@@ -0,0 +1,226 @@
+
+	return r.ReconcileHandler(adapter)
+}
+
+// getApplicationFromSnapshot loads from the cluster the Application referenced in the given ApplicationSnapshot.
+// If the ApplicationSnapshot doesn't specify an Component or this is not found in the cluster, an error will be returned.
+func (r *Reconciler) getApplicationFromSnapshot(context context.Context, snapshot *appstudioshared.ApplicationSnapshot) (*hasv1alpha1.Application, error) {
+	application := &hasv1alpha1.Application{}
+	err := r.Get(context, types.NamespacedName{
+		Namespace: snapshot.Namespace,
+		Name:      snapshot.Spec.Application,
+	}, application)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return application, nil
+}
+
+// getComponentFromSnapshot loads from the cluster the Component referenced in the given ApplicationSnapshot.
+// If the ApplicationSnapshot doesn't specify an Application or this is not found in the cluster, an error will be returned.
+func (r *Reconciler) getComponentFromSnapshot(context context.Context, snapshot *appstudioshared.ApplicationSnapshot) (*hasv1alpha1.Component, error) {
+	if componentLabel, ok := snapshot.Labels["component"]; ok {
+		component := &hasv1alpha1.Component{}
+		err := r.Get(context, types.NamespacedName{
+			Namespace: snapshot.Namespace,
+			Name:      componentLabel,
+		}, component)
+
+		if err != nil {
+			return nil, err
+		}
+
+		return component, nil
+	} else {
+		return nil, nil
+	}
+}
+
+// AdapterInterface is an interface defining all the operations that should be defined in an Integration adapter.
+type AdapterInterface interface {
+	EnsureAllReleasesExist() (results.OperationResult, error)
+	EnsureApplicationSnapshotEnvironmentBindingExist() (results.OperationResult, error)
+	EnsureAllIntegrationTestPipelinesExist() (results.OperationResult, error)
+}
+
+// ReconcileOperation defines the syntax of functions invoked by the ReconcileHandler
+type ReconcileOperation func() (results.OperationResult, error)
+
+// ReconcileHandler will invoke all the operations to be performed as part of an Integration reconcile, managing
+// the queue based on the operations' results.
+func (r *Reconciler) ReconcileHandler(adapter AdapterInterface) (ctrl.Result, error) {
+	operations := []ReconcileOperation{
+		adapter.EnsureAllReleasesExist,
+		adapter.EnsureApplicationSnapshotEnvironmentBindingExist,
+		adapter.EnsureAllIntegrationTestPipelinesExist,
+	}
+
+	for _, operation := range operations {
+		result, err := operation()
+		if err != nil || result.RequeueRequest {
+			return ctrl.Result{RequeueAfter: result.RequeueDelay}, err
+		}
+		if result.CancelRequest {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupController creates a new Integration reconciler and adds it to the Manager.
+func SetupController(manager ctrl.Manager, log *logr.Logger) error {
+	return setupControllerWithManager(manager, NewSnapshotReconciler(manager.GetClient(), log, manager.GetScheme()))
```

Comment:
I see the function has two versions, are we planning on using both of them?

Reply:
So each controller needs to have a `SetupController` function defined, and we call [all of them](https://github.com/redhat-appstudio/integration-service/pull/29/files#diff-46c1e02dfda1ba21d1ccabdf287921220f29dfdd6a75639fbf8a1a95bce01782R29-R30) from the `controllers.go` to set them all up.